client/opentracing/src/main/java/io/opentracing/APMSpan.java

```diff
@@ -84,6 +77,104 @@ protected void init(APMSpanBuilder builder, TraceReporter reporter) {
+            if (traceIds.size() == 1) {
+                if (builder.references.get(0).getReferredTo() instanceof APMSpan) {
+                    traceContext.initTraceState(((APMSpan)builder.references.get(0).getReferredTo()).getTraceContext());
+                } else if (builder.references.get(0).getReferredTo() instanceof APMSpanBuilder) {
+                    traceContext.initTraceState(((APMSpanBuilder)builder.references.get(0).getReferredTo()).getState());
+                }
+            }
+        }
+    
+        // Check if other references
+        for (Reference ref : builder.references) {
+            if (primaryRef == ref) {
+                continue;
+            }
+            // Setup correlation ids for other references
+            if (ref.getReferredTo() instanceof APMSpan) {
+                APMSpan referenced = (APMSpan) ref.getReferredTo();
+    
+                String nodeId = referenced.getNodePath();
+                getNodeBuilder().addCorrelationId(new CorrelationIdentifier(Scope.CausedBy, nodeId));
+
+            } else if (ref.getReferredTo() instanceof APMSpanBuilder
+                    && ((APMSpanBuilder)ref.getReferredTo()).getState().containsKey(Constants.HAWKULAR_APM_ID)) {
+                getNodeBuilder().addCorrelationId(new CorrelationIdentifier(Scope.Interaction,
+                        ((APMSpanBuilder)ref.getReferredTo()).getState().get(Constants.HAWKULAR_APM_ID).toString()));
+            }
+        }
+    }
+
+    /**
+     * This method identifies the primary 'parent' reference that should be used
+     * to link the associated span to an existing trace instance.
+     *
+     * @param references The list of references
+     * @return The primary reference, or null if one could not be determined
+     */
+    protected static Reference findPrimaryReference(List<Reference> references) {
+        List<Reference> followsFrom = references.stream()
+                .filter(ref -> ref.getReferenceType() == References.FOLLOWS_FROM)
+                .collect(Collectors.toList());
+
+        List<Reference> childOfSpan = references.stream()
+                .filter(ref -> ref.getReferenceType() == References.CHILD_OF
+                        && ref.getReferredTo() instanceof APMSpan)
+                .collect(Collectors.toList());
+
+        List<Reference> childOfSpanContext = references.stream()
+                .filter(ref -> ref.getReferenceType() == References.CHILD_OF
+                        && ref.getReferredTo() instanceof APMSpanBuilder)
+                .collect(Collectors.toList());
+
+        if (!childOfSpanContext.isEmpty()) {
+            if (childOfSpanContext.size() == 1) {
+                return childOfSpanContext.get(0);
+            }
+            return null;
+        }
+        if (!childOfSpan.isEmpty()) {
+            if (childOfSpan.size() == 1) {
+                return childOfSpan.get(0);
+            }
```

Comment:
There should be a null there. Will add.
The suggested code would return null if there was more than one followsFrom, even though only a single extractedContext - so doesn't represent the same logic.